### ISSUE TITLE
Using mocha parameter matchers w/stubbed functions?

### ISSUE THREAD

**Issue by wcooley** [Accuity/rspec-puppet-utils#3]
It does not seem possible to use mocha parameter matchers with parameters to stubbed functions. I've tried things like:

```
let!(:mock_hiera) do
  MockFunction.new('hiera') { |f|
    f.stub.with(['foo::bar', regexp_matches(/.*/)]).returns('whatever')
  }
end
```

resulting in:

```
     Mocha::ExpectationError:
       unexpected invocation: #<RSpecPuppetUtils::MockFunction:0x7f952255d660>.call(['foo::bar', 'actual second parameter']
     satisfied expectations:
       - allowed any number of times, not yet invoked: #<RSpecPuppetUtils::MockFunction:0x7f3c56464b70>.call(['openssl::origin_path', regexp_matches(/.*/)])
```

I am presuming (based on the lack of examples in the mocha documentation) that mocha does not deeply apply the parameter matchers, so if you pass an array it does not look at the elements of the array to see if they are parameter matchers?

If so, this makes them effectively useless for Puppet functions, since all of the parameters are passed as a single array. (Except for `any_parameters`, but that's unsatisfactory for obvious reasons.)

In this instance, I'd like to ignore the hiera default (the 2nd parameter) because it's dynamic -- it's based on the `get_module_path` function and the result of the hiera lookup is used as the parameter the `file` function, which requires an absolute path.

I'm working around it by stubbing `:get_module_path`; I could also get the fixtures path and dynamically work out what the default hiera value would be. In either case, this is a lot of ceremony to satisfy requirements for a dependency a couple of modules deep, and it requires me knowing how that dependency is building the default parameter for hiera, so it makes for a fragile coupling.

**Comment by TomPoulton** [comment]
Hey @wcooley and @riton, have a look at the readme in the `function-args` branch, I've tweaked the internals and added the `stubbed` and `expected` methods to replace the old `stub` and `expect` methods to help with parameter matching

I wanted to keep the old methods so that everything's backwards compatible, and I'm open to suggestions on the naming

**Comment by TomPoulton** [comment]
Looping in @jumanjiman as he's played around with the project

**Comment by TomPoulton** [comment]
This is fixed as of 2.1.0

The `#stubbed` and `#expected` methods take args instead of an array, so they can work with parameter matchers
